### DISCUSSION TITLE
Manual Device / Device-type Creation for AFL Hyperscale MPO Cassette

### DISCUSSION THREAD

**Discussion by kznfriks** [netbox-community/netbox#11975] [Q&A]
Good day, 

I was wondering if there is a reference to using AFL hyperscale MPO devices within Netbox , and preferably by manual creation as i am still exploring the product and most of the feeds are a bit overwhelming when getting to writing code to do imports or executions?

My datacenter makes use of  2 types of  AFL - Tray 's , 1 that  can house either 5 x AFL cassettes in 1 U format , Horizontally , or 1 that house 12 x AFL cassettes Vertically over 2 U rack space.

each AFL cassette has 2 x Rear ports  and 12 Front ports

the mapping of the MPO is as follow:

Rear-left (A) MPO (7-12) --mapped to Front Right (B) LC port 7 - 12
Rear-Right (B) MPO (1-6) --mapped to Front Left (A) LC port 1 - 6

Please note i generated A and B and as A-left B-Right

The AFL rear Ports is a single LC connector that can connect to another AFL Rear port  and thus have 6 x LC fibre connections between the 2 devices / cabinets. 
My installation currently calls for the rear A-ports to go to a Cabinet for storage connections and the B-ports to a different Cabinet for Network Connections.

Can someone please help in guiding me as to what should be a device , device type , module etc. ?

when trying to use an MPO device type on the rear port i cannot connect more than 2 front ports.

**Comment by candlerb** [comment]
> My datacenter makes use of 2 types of AFL - Tray 's , 1 that can house either 5 x AFL cassettes in 1 U format , Horizontally , or 1 that house 12 x AFL cassettes Vertically over 2 U rack space.

I used a system almost exactly like those: [Complete Connect MX Series](https://www.completeconnect.co.uk/mx-xpress-pre-terminated-modular-fibre/).

The first stage was to model the 1U and 2U frames as devices, with either 5 or 12 device bays.

The next stage was to model the cassettes themselves, as child devices.  These had 12 LC pairs on the front and some sort of MPO on the rear (not sure if it was 1 or 2 - the data centre installed them, and we connected whole cassettes back to back, so logically it was a single connection)

Since we don't have any simplex fibres, I modelled each LC pair as a single "frontport", and the MPO(s) on the back as a single "rearport".

> The AFL rear Ports is a single LC connector that can connect to another AFL Rear port and thus have 6 x LC fibre connections between the 2 devices / cabinets.

Is that a typo? I thought you said that the rear port is an MPO connector, not LC.

> My installation currently calls for the rear A-ports to go to a Cabinet for storage connections and the B-ports to a different Cabinet for Network Connections.

That sounds easy enough, and is only slightly more complex than mine. Build the Device Type for the cassette with:

* Two rear ports: say they are called "A" and "B" to use your current naming scheme
* Twelve front ports, say F[1-12] (or name them however you like)
* Associate F1-F6 to rear port "B" (positions 1 to 6) and F7-F12 to rear port "A" (positions 1 to 6). These associations are set at the time you create the frontports.

To create the cables between the rearports, I'd suggest a CSV import.  There are also scripts like [this one](https://github.com/netbox-community/reports/blob/master/scripts/multi_connect.py) but are aimed at when you have a whole bunch of separate rearports that need to be linked one by one.

(The Connect MX also has 4-way RJ45 modules, which are just RJ45 couplers, so there are 4 frontports and 4 rearports in the module)

**Comment by kznfriks** [comment]
Awesome , thank you very much for your response. this looks it will help me to get started. i will try it out and give you feedback!

Thanks Again!

**Reply by candlerb** [reply]
Great!

The other advice I'd give is: pick a naming scheme for the cassettes themselves which is consistent, and makes it obvious which ones are paired.

In my case it was simple because I wasn't splitting the rearports. My naming scheme was `mod-lc-<this_end>-<far_end>`. For example, if I had a link between rack S18 and rack R18, then the cassette in rack S18 was called "mod-lc-s18-r18" and the linked cassette in rack R18 was "mod-lc-r18-s18"
